src/draw/sw/lv_draw_sw_img.c

```diff
@@ -554,46 +554,77 @@ static void recolor(lv_area_t relative_area, uint8_t * src_buf, uint8_t * dest_b
             const uint8_t * src_buf_tmp = src_buf + src_stride * relative_area.y1 + relative_area.x1 * 2;
             int32_t img_stride_px = src_stride / 2;
 
-            uint16_t c_mult[3];
-            c_mult[0] = (color.blue >> 3) * mix;
-            c_mult[1] = (color.green >> 2) * mix;
-            c_mult[2] = (color.red >> 3) * mix;
             uint16_t * buf16_src = (uint16_t *)src_buf_tmp;
             uint16_t * buf16_dest = (uint16_t *)dest_buf;
-
-            int32_t y;
-            for(y = 0; y < h; y++) {
-                int32_t x;
-                for(x = 0; x < w; x++) {
-                    *buf16_dest = (((c_mult[2] + ((buf16_src[x] >> 11) & 0x1F) * mix_inv) << 3) & 0xF800) +
-                                  (((c_mult[1] + ((buf16_src[x] >> 5) & 0x3F) * mix_inv) >> 3) & 0x07E0) +
-                                  ((c_mult[0] + (buf16_src[x] & 0x1F) * mix_inv) >> 8);
-                    buf16_dest++;
+            uint16_t color16 = lv_color_to_u16(color);
+            if(mix >= LV_OPA_MAX) {
+                int32_t y;
+                for(y = 0; y < h; y++) {
+                    int32_t x;
+                    for(x = 0; x < w; x++) {
+                        *buf16_dest = color16;
+                        buf16_dest++;
+                    }
+                    buf16_src += img_stride_px;
+                }
+            }
+            else {
+                uint16_t c_mult[3];
+                c_mult[0] = (color.blue >> 3) * mix;
+                c_mult[1] = (color.green >> 2) * mix;
+                c_mult[2] = (color.red >> 3) * mix;
+
+                int32_t y;
+                for(y = 0; y < h; y++) {
+                    int32_t x;
+                    for(x = 0; x < w; x++) {
+                        *buf16_dest = (((c_mult[2] + ((buf16_src[x] >> 11) & 0x1F) * mix_inv) << 3) & 0xF800) +
+                                      (((c_mult[1] + ((buf16_src[x] >> 5) & 0x3F) * mix_inv) >> 3) & 0x07E0) +
+                                      ((c_mult[0] + (buf16_src[x] & 0x1F) * mix_inv) >> 8);
+                        buf16_dest++;
+                    }
+                    buf16_src += img_stride_px;
```

Comment:
is there guarantee a buffer overflow never occurs?

Reply:
@uLipe as these are running on the image width/height no under/overflow should happen